### ISSUE TITLE
TreeDataProvider: allow specifying a command to execute when a TreeItem is highlighted

### ISSUE THREAD

**Issue by patrys** [microsoft/vscode#34131]
This would allow extensions that use tree views for navigation match the behavior of the Quick Pick panel.

Related: https://github.com/patrys/vscode-code-outline/issues/5

**Comment by sandy081** [comment]
@patrys You can provide a command that will be executed when the tree item is selected. See here - https://github.com/microsoft/vscode/blob/master/src/vs/vscode.d.ts#L4523

Is not this what you intended to want?

**Comment by patrys** [comment]
That works fine but in this case I meant highlighted using the keyboard so pressing up/down would give you a preview of where you'd end up in code.

**Comment by sandy081** [comment]
We have two actions with keyboard - focus and selection. 
`up` and `down` focusses the item and `enter` selects the item. 
Commands are executed on selection

**Comment by patrys** [comment]
Sorry, let me rephrase that. I would like to be able to do some things when you use arrows (item focused) and another only when the item is confirmed (item selected). In my case it would be revealing parts of code on navigation and activating the editor once enter is pressed. I imagine it would be essential once we get search support for custom tree views.

**Comment by dakaraphi** [comment]
@sandy081 I am building an extension where this would greatly improve the experience.  That is to be able to provide a command for focus events.

Any chance to look at this again?

**Comment by sandy081** [comment]
Since, I started exposing events this can also be an event.

**Comment by dakaraphi** [comment]
> Since, I started exposing events this can also be an event.

Sounds good.  Also, are you suggesting this might get covered by other work `exposing events` you are already doing?

**Comment by sandy081** [comment]
@dakaraphi No... This milestone events for expand/collapse are added. It's just about the approach I wanted to take here.

**Comment by RandomFractals** [comment]
@alexr00 would be nice to have `onDidChangeFocus` event similar to `onDidChangeSelection` to utilize arrow keys for tree view nav and related 'outline' views display. My use case: https://github.com/RandomFractals/vscode-snippets-viewer/issues/56

**Comment by alexr00** [comment]
Duplicate of https://github.com/microsoft/vscode/issues/170248, since this issue would be resolved by the new API.
